File: examples/spring-boot/src/main/java/com/example/bookshop/Store.java
package com.example.bookshop;

import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.UnaryOperator;

/** The shop's shelves, in memory. */
@Component
public class Store {
    private final Map<String, Author> authors = Map.of(
        "a1", new Author("a1", "Ursula K. Le Guin"),
        "a2", new Author("a2", "Frank Herbert"));

    private final Map<String, Book> books = new ConcurrentHashMap<>(Map.of(
        "b1", new Book("b1", "A Wizard of Earthsea", 3, "a1", new BigDecimal("4.20")),
        "b2", new Book("b2", "The Left Hand of Darkness", 0, "a1", new BigDecimal("5.10")),
        "b3", new Book("b3", "Dune", 7, "a2", new BigDecimal("6.00"))));

    private final AtomicInteger authorLookups = new AtomicInteger();

    public Optional<Book> book(String id) {
        return Optional.ofNullable(books.get(id));
    }

    public List<Book> books() {
        return books.values().stream().sorted(Comparator.comparing(Book::id)).toList();
    }

    /** Many authors in one lookup, the way one {@code WHERE id IN (...)} query would fetch them. */
    public List<Author> authors(List<String> ids) {
        authorLookups.incrementAndGet();
        return ids.stream().map(authors::get).toList();
    }

    /** How many times authors were looked up, to see batching at work. */
    public int authorLookups() {
        return authorLookups.get();
    }

    /**
     * Replaces a book with what {@code change} makes of it, atomically, and returns the new copy. An exception thrown
     * by {@code change} leaves the book as it was.
     */
    public Optional<Book> update(String id, UnaryOperator<Book> change) {
        return Optional.ofNullable(books.computeIfPresent(id, (key, book) -> change.apply(book)));
    }
}

File: examples/kotlin/src/main/kotlin/com/example/bookshop/Store.kt
package com.example.bookshop

import java.math.BigDecimal
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

data class Author(val id: String, val name: String)

data class Book(val id: String, val title: String, val stock: Int, val authorId: String, val costPrice: BigDecimal)

/** The shop's shelves, in memory. Each server gets a fresh one. */
class Store {
    private val authors = listOf(
        Author("a1", "Ursula K. Le Guin"),
        Author("a2", "Frank Herbert"),
    ).associateBy { it.id }

    private val books = ConcurrentHashMap(
        listOf(
            Book("b1", "A Wizard of Earthsea", stock = 3, authorId = "a1", costPrice = BigDecimal("4.20")),
            Book("b2", "The Left Hand of Darkness", stock = 0, authorId = "a1", costPrice = BigDecimal("5.10")),
            Book("b3", "Dune", stock = 7, authorId = "a2", costPrice = BigDecimal("6.00")),
        ).associateBy { it.id },
    )

    /** How many times authors were looked up, to see batching at work. */
    val authorLookups = AtomicInteger()

    fun book(id: String): Book? = books[id]

    fun books(): List<Book> = books.values.sortedBy { it.id }

    /** Many authors in one lookup, the way one `WHERE id IN (...)` query would fetch them. */
    fun authors(ids: List<String>): List<Author?> {
        authorLookups.incrementAndGet()
        return ids.map { authors[it] }
    }

    /**
     * Replaces a book with what [change] makes of it, atomically, and returns the new copy; null when there is no such
     * book. An exception thrown by [change] leaves the book as it was.
     */
    fun update(id: String, change: (Book) -> Book): Book? = books.computeIfPresent(id) { _, book -> change(book) }
}
